File: airbyte-commons/src/test/java/io/airbyte/commons/lang/ExceptionsTest.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.commons.lang;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

@SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")
class ExceptionsTest {

  @Test
  void testToRuntime() {
    assertEquals("hello", Exceptions.toRuntime(() -> callable("hello", false)));
    assertThrows(RuntimeException.class, () -> Exceptions.toRuntime(() -> callable("goodbye", true)));
  }

  @Test
  void testToRuntimeVoid() {
    final List<String> list = new ArrayList<>();
    assertThrows(RuntimeException.class, () -> Exceptions.toRuntime(() -> voidCallable(list, "hello", true)));
    assertEquals(0, list.size());

    Exceptions.toRuntime(() -> voidCallable(list, "goodbye", false));
    assertEquals(1, list.size());
    assertEquals("goodbye", list.get(0));
  }

  @Test
  void testSwallow() {
    Exceptions.swallow(() -> {
      throw new RuntimeException();
    });
    Exceptions.swallow(() -> {
      throw new Exception();
    });
  }

  private String callable(final String input, final boolean shouldThrow) throws IOException {
    if (shouldThrow) {
      throw new IOException();
    } else {
      return input;
    }
  }

  private void voidCallable(final List<String> list, final String input, final boolean shouldThrow) throws IOException {
    if (shouldThrow) {
      throw new IOException();
    } else {
      list.add(input);
    }
  }

}

File: airbyte-cdk/java/airbyte-cdk/dependencies/src/test/kotlin/io/airbyte/commons/lang/ExceptionsTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.commons.lang

import java.io.IOException
import java.util.concurrent.Callable
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class ExceptionsTest {
    @Test
    fun testToRuntime() {
        Assertions.assertEquals("hello", Exceptions.toRuntime<String> { callable("hello", false) })
        Assertions.assertThrows(RuntimeException::class.java) {
            Exceptions.toRuntime(Callable { callable("goodbye", true) })
        }
    }

    @Test
    fun testToRuntimeVoid() {
        val list: MutableList<String> = ArrayList()
        Assertions.assertThrows(RuntimeException::class.java) {
            Exceptions.toRuntime { voidCallable(list, "hello", true) }
        }
        Assertions.assertEquals(0, list.size)

        Exceptions.toRuntime { voidCallable(list, "goodbye", false) }
        Assertions.assertEquals(1, list.size)
        Assertions.assertEquals("goodbye", list[0])
    }

    @Test
    fun testSwallow() {
        Exceptions.swallow { throw RuntimeException() }
        Exceptions.swallow { throw Exception() }
    }

    @Throws(IOException::class)
    private fun callable(input: String, shouldThrow: Boolean): String {
        if (shouldThrow) {
            throw IOException()
        } else {
            return input
        }
    }

    @Throws(IOException::class)
    private fun voidCallable(list: MutableList<String>, input: String, shouldThrow: Boolean) {
        if (shouldThrow) {
            throw IOException()
        } else {
            list.add(input)
        }
    }
}
